tests/run/critical_sections.pyx

```diff
@@ -0,0 +1,121 @@
+# mode: run
+
+cimport cython
+
+import threading
+
+# This test is most useful on free-threading builds.
+# It should pass on regular builds just by virtue of
+# the GIL.
+
+def test_single_critical_section(n_threads, n_loops):
+    """
+    >>> test_single_critical_section(4, 100)
+    """
+    cdef int count = 0
+    lock = object()
+    barrier = threading.Barrier(n_threads)
+
+    def worker():
+        nonlocal count
+        barrier.wait()
+        for i in range(n_loops):
+            with cython.critical_section(lock):
+                count += i
+
+    threads = [
+        threading.Thread(target=worker)
+        for _ in range(n_threads)
+    ]
+    for t in threads:
+        t.start()
+    for t in threads:
+        t.join()
+
+    expected = ((n_loops * (n_loops - 1))/2)*n_threads
+    assert count == expected, (count, expected)
+
+def test_double_critical_section(n_loops):
+    """
+    >>> test_double_critical_section(100)
+    """
+    cdef int a = 2
+    cdef int b = 3
+    barrier = threading.Barrier(3)
+    finished_event = threading.Event()
+    failed = threading.Event()
+    lock_a = object()
+    lock_b = object()
+
+    def reader1():
+        barrier.wait()
+        while not finished_event.is_set():
+            for _ in range(10):
+                with cython.critical_section(lock_a):
+                    if not (a==2 or a==3):
```

Comment:
I actually only wanted to add spaces around the comparison operators, but `not in` seems more pythonic here.
```suggestion
                    if a not in (2, 3):
```